### ISSUE TITLE
extends not working for generics as expected

### ISSUE THREAD

**Issue by aboyton** [Microsoft/TypeScript#20840]
**TypeScript Version:**  2.6.2

**Code**
[Playground link](https://www.typescriptlang.org/play/#src=function%20function1%3CT%20extends%20'a'%20%7C%20'b'%3E(key%3A%20T)%20%7B%0D%0A%20%20switch%20(key)%20%7B%0D%0A%20%20%20%20case%20'a'%3A%20%7B%0D%0A%20%20%20%20%20%20console.log(key)%3B%0D%0A%20%20%20%20%20%20return%20%5B%5D%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20default%3A%0D%0A%20%20%20%20%20%20console.log(key)%3B%0D%0A%20%20%20%20%20%20return%20%5B%5D%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20function2(key%3A%20'a'%20%7C%20'b')%20%7B%0D%0A%20%20switch%20(key)%20%7B%0D%0A%20%20%20%20case%20'a'%3A%20%7B%0D%0A%20%20%20%20%20%20console.log(key)%3B%0D%0A%20%20%20%20%20%20return%20%5B%5D%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20default%3A%0D%0A%20%20%20%20%20%20console.log(key)%3B%0D%0A%20%20%20%20%20%20return%20%5B%5D%3B%0D%0A%20%20%7D%0D%0A%7D)

```ts
function function1<T extends 'a' | 'b'>(key: T) {
  switch (key) {
    case 'a': {
      console.log(key);
      return [];
    }
    default:
      console.log(key);
      return [];
  }
}
```

```ts
function function2(key: 'a' | 'b') {
  switch (key) {
    case 'a': {
      console.log(key);
      return [];
    }
    default:
      console.log(key);
      return [];
  }
}
```

(Obviously this is a minimal example, and in real work code I'd have multiple parameters using the type `T` in their types.)

**Expected behavior:**

I would expect that in either case `key` is of type `'a' | 'b'`.

**Actual behavior:**
 In actuality in the first example (`function1 `) I have `key: T extends "a" | "b"` and inside the `case 'a'` branch `key` is of type `never`. 

Alternatively maybe I'm using `extends` wrong, but I can't seem to find a way of specifying that `T` is either `'a'` or `'b'`. For example if I have `function function1<T extends 'a' | 'b'>(key1: T, key2: T) {` then I'd like to be able to specify that `key1` and `key2` are of the same type (either both `'a'` or both `'b'`).

**Comment by DanielRosenwasser** [comment]
It seems extremely strange, at the least, that in the first function, `key` gets narrowed to `never`.

**Comment by ahejlsberg** [comment]
Yes, definitely not right that `key` is narrowed to `never`. The following should not error:

```ts
function function1<T extends 'a' | 'b'>(key: T) {
  switch (key) {
    case 'a':
      key.toLowerCase();  // Error, but shouldn't be
      break;
    default:
      key.toLowerCase();
      break;
  }
}
```

**Comment by ahejlsberg** [comment]
With the fix in #20882 the type of `key` is no longer narrowed in the `switch` statement. That is the intended behavior and matches what happens if the same logic is written as `if (key === 'a')`. We currently don't narrow constrained type parameters and thus, even after the fix, `key` isn't narrowed to type `'a'` in the first example.

When a type parameter `T` includes a unit type `U` in its constraint, we could consider making `U` assignable to `T` and allow narrowing of `T` to `U`. However, that is larger change beyond the scope of this issue.
